src/Orleans.Runtime/Streams/PersistentStream/PersistentStreamPullingAgent.cs

```diff
@@ -352,16 +353,20 @@ public void RemoveSubscriber_Impl(GuidId subscriptionId, StreamId streamId)
                     int maxCacheAddCount = queueCache?.GetMaxAddCount() ?? QueueAdapterConstants.UNLIMITED_GET_QUEUE_MSG;
                     if (maxCacheAddCount != QueueAdapterConstants.UNLIMITED_GET_QUEUE_MSG && maxCacheAddCount <= 0)
                         return;
-
+                    
+                    var retryMax = 6;
                     // If read succeeds and there is more data, we continue reading.
                     // If read succeeds and there is no more data, we break out of loop
-                    // If read fails, we try again, with backoff policy.
-                    //    This prevents spamming backend queue which may be encountering transient errors.
-                    //    We retry until the operation succeeds or we are shutdown.
+                    // If read fails, we retry 6 more times, with backoff policy.
+                    //    we log each failure as warnings. After 6 times retry if still fail, we break out of loop and log an error
                     bool moreData = await AsyncExecutorWithRetries.ExecuteWithRetries(
-                        i => ReadFromQueue((QueueId)state, receiver, maxCacheAddCount),
-                        AsyncExecutorWithRetries.INFINITE_RETRIES,
-                        (e, i) => !IsShutdown,
+                        i => ReadFromQueue(queueId, receiver, maxCacheAddCount),
+                        retryMax,
+                        (e, i) =>
+                        {
+                            this.logger.Warn(ErrorCode.PersistentStreamPullingAgent_12, $"Exception while retrying the {i}th time reading from queue {queueId}. Exception is {e}");
+                            return !IsShutdown;
```

Comment:
this was fine as a lambda when it was a single operation, more complex logic (like this) should probably be in it's own function.